# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

- **Consider a significance level of 5% for all tests.**

In [1]:
# import numpy and pandas
import pandas as pd
import numpy as np
import scipy.stats as stats


# Challenge 1 - Independent Sample T-tests

In this challenge, we will be using the Pokemon dataset. Before applying statistical methods to this data, let's first examine the data.

To load the data, run the code below.

In [2]:
# Run this code:

pokemon = pd.read_csv('pokemon.csv')

Let's start off by looking at the `head` function in the cell below.

In [3]:
pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

The first thing we would like to do is compare the legendary Pokemon to the regular Pokemon. To do this, we should examine the data further. What is the count of legendary vs. non legendary Pokemons?

In [4]:
pokemon["Legendary"].value_counts()

Legendary
False    735
True      65
Name: count, dtype: int64

Compute the mean and standard deviation of the total points for both legendary and non-legendary Pokemon.

In [5]:
leg = pokemon[pokemon["Legendary"] == True]
norm = pokemon[pokemon["Legendary"] == False]

print(f"legendary mean = {round(leg.Total.mean(), 2)}")
print(f"legendary std = {round(leg.Total.std(), 2)}")
print(f"non-legendary mean = {round(norm.Total.mean(), 2)}")
print(f"non-legendary std = {round(norm.Total.std(), 2)}")

legendary mean = 637.38
legendary std = 60.94
non-legendary mean = 417.21
non-legendary std = 106.76


The computation of the mean might give us a clue regarding how the statistical test may turn out; However, it certainly does not prove whether there is a significant difference between the two groups.

In the cell below, use the `ttest_ind` function in `scipy.stats` to compare the the total points for legendary and non-legendary Pokemon. Since we do not have any information about the population, assume the variances are not equal.

In [6]:
t_stat, p_value = stats.ttest_ind(leg.Total, norm.Total, equal_var=False)
p_value

9.357954335957446e-47

What do you conclude from this test? Write your conclusions below.

very low p value, this means the null hypothesis is rejected. the legendary status has a correlation with the total points

How about we try to compare the different types of pokemon? In the cell below, list the types of Pokemon from column `Type 1` and the count of each type.

In [7]:
pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

In [8]:
table_1 = pokemon.groupby("Type 1").size().reset_index().sort_values(0, ascending=False)

In [9]:
table_1.head()

Type 1    0
17    Water  112
12   Normal   98
9     Grass   70
0       Bug   69
14  Psychic   57

Since water is the largest group of Pokemon, compare the mean and standard deviation of water Pokemon to all other Pokemon.

In [10]:
# water
water_mean, water_std = pokemon.groupby("Type 1").get_group("Water")["Total"].agg(["mean", "std"]).round()
print(water_mean, water_std)

430.0 113.0


In [11]:
water = pokemon.groupby("Type 1").get_group("Water")["Total"]

In [12]:
# all other pokemon
rest = pokemon[pokemon["Type 1"] != "Water"]
rest_total = pokemon[pokemon["Type 1"] != "Water"]["Total"]

rest_mean, rest_std = rest["Total"].agg(["mean", "std"]).round()
print(rest_mean, rest_std)

436.0 121.0


Perform a hypothesis test comparing the mean of total points for water Pokemon to all non-water Pokemon. Assume the variances are equal. 

In [13]:
t_stat, p_value = stats.ttest_ind(water, rest_total, equal_var=True)
p_value

0.6587140317488793

Write your conclusion below.

there seems to be no big difference of the mean values of water pokemon and all others. this can also be seen in the means, water = 430, rest = 436

# Challenge 2 - Matched Pairs Test

In this challenge we will compare dependent samples of data describing our Pokemon. Our goal is to see whether there is a significant difference between each Pokemon's defense and attack scores. Our hypothesis is that the defense and attack scores are equal. In the cell below, import the `ttest_rel` function from `scipy.stats` and compare the two columns to see if there is a statistically significant difference between them.

In [14]:
pokemon

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 13 columns]

In [15]:
from scipy.stats import ttest_rel
t_stat, p_value = stats.ttest_rel(pokemon["Attack"], pokemon["Defense"])
p_value

1.7140303479358558e-05

Describe the results of the test in the cell below.

very small p value, this means we reject null hypothesis. there is a statistically significant difference between them.

We are also curious about whether therer is a significant difference between the mean of special defense and the mean of special attack. Perform the hypothesis test in the cell below. 

In [16]:
t_stat, p_value = stats.ttest_rel(pokemon["Sp. Atk"], pokemon["Sp. Def"])
p_value

0.3933685997548122

Describe the results of the test in the cell below.

not so much here. the value is bigger than 0.05, this means we can not reject null hypothesis. there is no big statistical difference.

As you may recall, a two sample matched pairs test can also be expressed as a one sample test of the difference between the two dependent columns.

Import the `ttest_1samp` function and perform a one sample t-test of the difference between defense and attack. Test the hypothesis that the difference between the means is zero. Confirm that the results of the test are the same.

In [17]:
pokemon["diff"] = pokemon["Defense"] - pokemon["Attack"]
pokemon.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  diff  
0       65       65     45           1      False     0  
1       80       80     60           1      False     1  
2      100      100     80           1      False     1  
3      122      120     80           1      False    23  
4       60       50     65           1      False    -9

In [18]:
from scipy.stats import ttest_1samp
t_stat, p_value = stats.ttest_1samp(pokemon["diff"], 0)
p_value

1.7140303479358558e-05

getting the same as above

# Bonus Challenge - The Chi-Square Test

The Chi-Square test is used to determine whether there is a statistically significant difference in frequencies. In other words, we are testing whether there is a relationship between categorical variables or rather when the variables are independent. This test is an alternative to Fisher's exact test and is used in scenarios where sample sizes are larger. However, with a large enough sample size, both tests produce similar results. Read more about the Chi Squared test [here](https://en.wikipedia.org/wiki/Chi-squared_test).

In the cell below, create a contingency table using `pd.crosstab` comparing whether a Pokemon is legenadary or not and whether the Type 1 of a Pokemon is water or not.

In [29]:
legend = pokemon["Legendary"]==True
type1 = pokemon["Type 1"]=="Water"

In [39]:
my_crosstab = pd.crosstab(legend, type1, rownames=["legendary"], colnames=["water"], margins=True)
my_crosstab

water      False  True  All
legendary                  
False        627   108  735
True          61     4   65
All          688   112  800

Perform a chi-squared test using the `chi2_contingency` function in `scipy.stats`. You can read the documentation of the function [here](https://docs.scipy.org/doc/scipy-0.15.1/reference/generated/scipy.stats.chi2_contingency.html).

In [42]:
from scipy.stats import chi2_contingency
chi2_stats, chi2_pvalue, _, _ = chi2_contingency(my_crosstab)
chi2_pvalue

0.46024514032769426

Based on a 95% confidence, should we reject the null hypothesis?

based on this number (pvalue = 0.46) we should not reject the null hypothesis. the legendary ratio seems to be equally distributed over the water and the non-water types.